File: 113_Path Sum II.java
// Time Complexity: O(n^2)
// Space Complexity: O(n)

/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    private List<List<Integer>> res;
    private List<Integer> curPath;

    private void bt(TreeNode node, int target) {
        if (node == null) return;

        curPath.add(node.val);
        target -= node.val;

        if (node.left == null && node.right == null) {
            if (target == 0) res.add(new ArrayList<>(curPath));
            curPath.remove(curPath.size()-1);
            return;
        }

        bt(node.left, target);
        bt(node.right, target);
        curPath.remove(curPath.size()-1);
    }

    public List<List<Integer>> pathSum(TreeNode root, int targetSum) {
        this.res = new ArrayList<>();
        this.curPath = new ArrayList<>();

        bt(root, targetSum);
        return this.res;
    }
}

File: 113_Path Sum II.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def pathSum(self, root: Optional[TreeNode], targetSum: int) -> List[List[int]]:
        def bt(node: Optional[TreeNode], target):
            if node == None: return

            curPath.append(node.val)
            target -= node.val

            if node.left == None and node.right == None:
                if target == 0: res.append(list(curPath))
                curPath.pop()
                return
            
            bt(node.left, target)
            bt(node.right, target)
            curPath.pop()

        res, curPath = [], []
        bt(root, targetSum)
        return res